### DISCUSSION TITLE
For Python API, what is the difference between duckdb.sql and relation.query?

### DISCUSSION THREAD

**Discussion by loneguardian** [duckdb/duckdb#8125] [Q&A] [answered]
I realise there are a few ways to do sql query in duckdb but I wasn't able to find much information on the differences between `duckdb.sql()` and `relation.query()`.

Is `relation.df()` the same as `fetch` but instead of Python type, it is exported into dataframes?

**Comment by Tishj** [comment, marked as answer]
duckdb.sql is actually duckdb.default_connection.sql
(it uses the default connection)

query, sql, from_query are aliases
The main difference with relation.query is that the relation is registered under the alias provided to the function

**Comment by Tishj** [comment]
Yes both `fetchall` and `df` both fully consume the result
The main difference being that `df` creates a pandas DataFrame and `fetchall` creates native python objects
